File: java/Components/Strata.Domain.Core/src/strata/domain/core/unitofwork/AbstractUnitOfWork.java
// ##########################################################################
// # File Name:	AbstractUnitOfWork.java
// #
// # Copyright:	2017, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataPersistence Framework.
// #
// #   			The StrataPersistence Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataPersistence Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataPersistence
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.domain.core.unitofwork;

import strata.domain.core.namedquery.INamedQuery;

import java.io.Serializable;
import java.util.Optional;
import java.util.Stack;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public abstract 
class AbstractUnitOfWork
    implements IUnitOfWork
{
    private IUnitOfWorkProvider itsProvider;
    private IUnitOfWorkState    itsCurrent;
    private Stack<Runnable> itsRollbackActions;
    
    /************************************************************************
     * Creates a new {@code AbstractUnitOfWork}. 
     *
     */
    protected 
    AbstractUnitOfWork(IUnitOfWorkProvider provider)
    {
        itsProvider = provider;
        itsCurrent  = ActiveUnitOfWorkState.getInstance();
        itsRollbackActions = new Stack<>();
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<E>
    insertNew(Class<K> keyType,Class<E> entityType,E newEntity)
    {
        return itsCurrent.insertNew( this,keyType,entityType,newEntity );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<E>
    updateExisting(Class<K> keyType,Class<E> entityType,E existingEntity)
    {
        return
            itsCurrent
                .updateExisting( this,keyType,entityType,existingEntity );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<Void>
    removeExisting(Class<K> keyType,Class<E> entityType,E existingEntity)
    {
        System.out.println("AbstractUnitOfWOrk.removeExisting: " + Thread.currentThread().getName());
        return
            itsCurrent.removeExisting( this,keyType,entityType,existingEntity );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public IUnitOfWorkProvider 
    getProvider()
    {
        return itsProvider;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<Optional<E>>
    getExisting(Class<E> type,K key)
    {
        System.out.println("AbstractUnitOfWOrk.getExisting: " + Thread.currentThread().getName());
        return itsCurrent.getExisting( this,type,key );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <E> CompletionStage<Optional<INamedQuery<E>>>
    getNamedQuery(Class<E> type,String queryName)
    {
        System.out.println("AbstractUnitOfWOrk.getNamedQuery: " + Thread.currentThread().getName());
        return itsCurrent.getNamedQuery( this,type,queryName );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<Boolean>
    hasExisting(Class<E> type,K key)
    {
        System.out.println("AbstractUnitOfWOrk.hasExisting: " + Thread.currentThread().getName());
        return itsCurrent.hasExisting( this,type,key );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <E> CompletionStage<Boolean>
    hasNamedQuery(Class<E> type,String queryName)
    {
        System.out.println("AbstractUnitOfWOrk.hasNamedQuery: " + Thread.currentThread().getName());
        return itsCurrent.hasNamedQuery( this,type,queryName );
    }

    /************************************************************************
     * {@inheritDoc}
     */
    @Override
    public void
    pushRollbackAction(Runnable action)
    {
        itsRollbackActions.push(action);
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isActive()
    {
        return itsCurrent.isActive( this );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isFailed()
    {
        return itsCurrent.isFailed( this );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isCommitted()
    {
        return itsCurrent.isCommitted( this );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isRolledback()
    {
        return itsCurrent.isRolledBack( this );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public CompletionStage<Void>
    commit() 
    {
        System.out.println("AbstractUnitOfWOrk.commit: " + Thread.currentThread().getName());
        return itsCurrent.commit( this );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public CompletionStage<Void>
    rollback()
    {
        System.out.println("AbstractUnitOfWOrk.rollback: " + Thread.currentThread().getName());
        return itsCurrent.rollback( this );
    }
    
    /************************************************************************
     *  
     *
     * @param state
     */
    protected void
    changeState(IUnitOfWorkState state)
    {
        itsCurrent = state;
    }

    /************************************************************************
     *  
     *
     * @param keyType TODO
     * @param entityType TODO
     * @param newEntity
     * @return
     */
    protected abstract <K extends Serializable,E> CompletionStage<E>
    doInsertNew(Class<K> keyType,Class<E> entityType,E newEntity);

    /************************************************************************
     *  
     *
     * @param keyType TODO
     * @param entityType TODO
     * @param existingEntity
     * @return
     */
    protected abstract <K extends Serializable,E> CompletionStage<E>
    doUpdateExisting(Class<K> keyType,Class<E> entityType,E existingEntity);

    /************************************************************************
     *  
     *
     * @param keyType TODO
     * @param entityType TODO
     * @param existingEntity
     */
    protected abstract <K extends Serializable,E> CompletionStage<Void>
    doRemoveExisting(Class<K> keyType,Class<E> entityType,E existingEntity);

    /************************************************************************
     *  
     *
     * @param type
     * @param key
     * @return
     */
    protected abstract <K extends Serializable,E> CompletionStage<Optional<E>>
    doGetExisting(Class<E> type,K key);

    /************************************************************************
     *  
     *
     * @param type
     * @param queryName
     * @return
     */
    protected abstract <E> CompletionStage<Optional<INamedQuery<E>>>
    doGetNamedQuery(Class<E> type,String queryName);

    /************************************************************************
     *  
     *
     * @param type TODO
     * @param key
     * @return
     */
    protected abstract <K extends Serializable,E> CompletionStage<Boolean>
    doHasExisting(Class<E> type,K key);

    /************************************************************************
     *  
     *
     * @param type
     * @param queryName
     * @return
     */
    protected abstract <E> CompletionStage<Boolean>
    doHasNamedQuery(Class<E> type,String queryName);
    
    /************************************************************************
     *  
     *
     */
    protected abstract CompletionStage<Void>
    doCommit();
    
    /************************************************************************
     *  
     *
     */
    protected CompletionStage<Void>
    doRollback()
    {
        return
            CompletableFuture.supplyAsync(
                () ->
                {
                    System.out.println("AbstractUnitOfWork.doRollback");
                    RollbackFailedException failed = null;

                    while (!itsRollbackActions.isEmpty())
                    {
                        try
                        {
                            Runnable action = itsRollbackActions.pop();

                            action.run();
                        }
                        catch (Exception e)
                        {
                            if (failed == null)
                                failed =
                                    new RollbackFailedException(
                                        "Rollback failed. See causes.");

                            failed.addCause(e);
                        }
                    }

                    if (failed != null)
                        throw new CompletionException(failed);

                    return null;
                });
    }
}

// ##########################################################################

File: csharp/Components/Strata.Domain.Core/UnitOfWork/AbstractUnitOfWork.cs
//  ##########################################################################
//  # File Name: AbstractUnitOfWork.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

using Strata.Domain.Core.NamedQuery;
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Strata.Foundation.Core.Utility;
using TaskFactory = Strata.Foundation.Core.Utility.TaskFactory;

namespace Strata.Domain.Core.UnitOfWork
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    ///  
    public abstract
    class AbstractUnitOfWork:
        IUnitOfWork
    {
        public IUnitOfWorkProvider     Provider { get; protected set; }
        internal IUnitOfWorkState      State { get; set; }
        internal Stack<RollbackAction> RollbackActions { get; set; }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>AbstractUnitOfWork</c> instance.
        /// </summary>
        /// 
        protected
        AbstractUnitOfWork(IUnitOfWorkProvider provider)
        {
            Provider        = provider;
            State           = ActiveUnitOfWorkState.GetInstance();
            RollbackActions = new Stack<RollbackAction>();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public abstract void 
        Dispose();

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<E> 
        Insert<K,E>(E entity) 
            where E: class
        {
            return State.Insert<K,E>(this,entity);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<E> 
        Update<K,E>(E entity) 
            where E: class
        {
            return State.Update<K,E>(this,entity);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task
        Remove<K,E>(E entity) 
            where E: class
        {
            return State.Remove<K,E>(this,entity);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<Optional<E>> 
        GetUnique<K,E>(K key) 
            where E: class
        {
            return State.GetUnique<K,E>(this,key);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<Optional<E>> 
        GetUniqueMatching<E>(Expression<Func<E,bool>> predicate) 
            where E: class
        {
            return State.GetUniqueMatching(this,predicate);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<IList<E>> 
        GetMatching<E>(Expression<Func<E,bool>> predicate) 
            where E: class
        {
            return State.GetMatching(this,predicate);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<IList<E>> 
        GetAll<E>() 
            where E: class
        {
            return State.GetAll<E>(this);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<INamedQuery<E>> 
        GetNamedQuery<E>(string queryName) 
            where E: class
        {
            return State.GetNamedQuery<E>(this,queryName);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<bool> 
        HasUnique<K,E>(K key) 
            where E: class
        {
            return State.HasUnique<K,E>(this,key);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<bool> 
        HasUniqueMatching<E>(Expression<Func<E,bool>> predicate) 
            where E: class
        {
            return State.HasUniqueMatching(this,predicate);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task<bool> 
        HasNamedQuery<E>(string queryName)
            where E: class
        {
            return State.HasNamedQuery<E>(this,queryName);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public void
        PushRollbackAction(RollbackAction rollbackAction)
        {
            RollbackActions.Push(rollbackAction);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public bool 
        IsActive()
        {
            return State.IsActive(this);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public bool 
        IsFailed()
        {
            return State.IsFailed(this);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public bool 
        IsCommitted()
        {
            return State.IsCommitted(this);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public bool 
        IsRolledBack()
        {
            return State.IsRolledBack(this);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task 
        Commit()
        {
            return State.Commit(this);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public Task 
        Rollback()
        {
            return State.Rollback(this);
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>/>
        /// <param name="entity">entity being inserted</param>
        /// <returns>the inserted entity</returns>
        /// 
        public abstract Task<E> 
        DoInsert<K,E>(E entity)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>/>
        /// <param name="entity">entity being updated</param>
        /// <returns>the updated entity</returns>
        /// 
        public abstract Task<E> 
        DoUpdate<K,E>(E entity)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>/>
        /// <param name="entity">entity being removed</param>
        public abstract Task
        DoRemove<K,E>(E entity)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>/>
        /// <param name="key">an entity's key</param>
        /// <returns>the entity if it exists or null</returns>
        /// 
        public abstract Task<Optional<E>> 
        DoGetUnique<K,E>(K key)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Retrieves the entity that matches the specified predicate. 
        /// </summary>
        /// <param name="predicate">
        ///     a predicate that matches a unique entity
        /// </param>
        /// <returns>the uniquely matching entity or null</returns>
        /// 
        public abstract Task<Optional<E>> 
        DoGetUniqueMatching<E>(Expression<Func<E,bool>> predicate)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  
        /// </summary>
        /// <param name="predicate"></param>
        /// 
        public abstract Task<IList<E>> 
        DoGetMatching<E>(Expression<Func<E,bool>> predicate)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  
        /// </summary>
        public abstract Task<IList<E>> 
        DoGetAll<E>()
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  
        /// </summary>
        /// <param name="queryName">query's name</param>
        /// <returns>query or null</returns>
        /// 
        public abstract Task<INamedQuery<E>> 
        DoGetNamedQuery<E>(string queryName)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  
        /// </summary>
        /// <param name="key"></param>
        /// 
        public abstract Task<bool> 
        DoHasUnique<K,E>(K key)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  
        /// </summary>
        /// <param name="predicate"></param>
        /// 
        public abstract Task<bool> 
        DoHasMatching<E>(Expression<Func<E,bool>> predicate)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  
        /// </summary>
        /// <param name="queryName">query's name</param>
        /// <returns>true if query exists, false otherwise</returns>
        /// 
        public abstract Task<bool> 
        DoHasNamedQuery<E>(string queryName)
            where E: class;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  
        /// </summary>
        /// 
        public abstract Task 
        DoCommit();

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  
        /// </summary>
        /// 
        public virtual Task 
        DoRollback()
        {
            Task task = 
                new Task(
                    () =>
                    {
                        RollbackFailedException failed = null;

                        while (RollbackActions.Count != 0)
                        {
                            try
                            {
                                RollbackAction action = RollbackActions.Pop();

                                action.Invoke();
                            }
                            catch (Exception e)
                            {
                                if (failed == null)
                                    failed =
                                        new RollbackFailedException(
                                            "Rollback failed. See causes.");

                                failed.Causes.Add(e);
                            }
                        }

                        if (failed != null)
                            throw failed;
                    });

            task.Start();
            return task;
        }
    }
}

//  ##########################################################################
